### PR TITLE
[fix](storage) Disable compaction before schema change is actually executed(#9032)

### LINKED ISSUE #9032: [Bug] Compaction leads to schema change error
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Version

0.14

### What's Wrong?
After #6365, doris may do comapction when a tablet in altering.

base tablet：[0-62][63-64][65-65][66-66][67-67]
new tablet:[50-65][66-66][67-67]
begin compaction in new tablet: [50-65][66-66]
begin schema change converting: `begin to remove all data from new tablet to prevent rewrite.`
new tablet: [50-65][66-66][67-67] will move to `TimestampedVersionTracker._stale_version_path_map`
compaction end in new tablet: [50-66] will be inserted into new tablet.
begin convert base tablet to new tablet:
insert [0-62] to new tablet and new tablet will be[0-62][50-66]
insert [63-64] will fail.
```version already exist, version revert occurred. tablet=XXXXX, version='63-64```
[65-65] and [66-66] will fail too.
```version already exist, version revert occurred. tablet=XXXXX, version='66-66```
insert [67-67] to new tablet, and new tabelt will be: [0-62][50-66][67]
Because the removed rowset can be query now, `check_version_integrity([0-67])` will success and the schema cahnge will success.
```
I0412 21:42:58.683157 60051 schema_change.cpp:1953] finish converting rowsets for new_tablet from base_tablet. base_tablet=XXXXX, new_tablet=XXXXX
I0412 21:42:58.700071 60051 engine_alter_tablet_task.cpp:52] success to create new alter tablet. res=0 base_tablet_id=XXXX, base_schema_hashXXXX, new_tablet_id=XXXX, new_schema_hash=XXXX
```

### What You Expected?

1. compaction should not add output rowset to tablet if the origin rowsets not in tablet's `_rs_version_map`.
2. schema change should check new tablet's `_rs_version_map`
3. tablet report should check if a tablet has intersected versions.

### How to Reproduce?

_No response_

### Anything Else?

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by liutang123**
# Proposed changes

Issue Number: close #9032

## Problem Summary:
As in issue, the combination and schema change at the same time may lead to version intersection.
Describe the overview of changes.
1. Do not do compaction before schema change is actually executed.
2. Set tablet as bad when it has version intersection.
3. Do not do schema change when it can not find appropriate versions to delete in new tablet.
4. Do not change rowsets after compaction if the rowsets of the tablet has changed.

## Checklist(Required)

1. Does it affect the original behavior: (Yes)
5. Has unit tests been added: (Yes/No/No Need)
6. Has document been added or modified: (Yes/No/No Need)
7. Does it need to update dependencies: (Yes/No)
8. Are there any changes that cannot be rolled back: (Yes/No)

### CODE REVIEW

File: be/src/olap/tablet.cpp [hunk 1]
```diff
@@ -275,6 +277,23 @@ void Tablet::modify_rowsets(std::vector<RowsetSharedPtr>& to_add,
         same_version = false;
     }
 
+    if (check_delete) {
+        for (auto& rs : to_delete) {
+            auto find_rs = _rs_version_map.find(rs->version());
```
Comment (yangzhg) [inline be/src/olap/tablet.cpp:282]: maybe need to lock `_meta_lock`

Comment (liutang123) [reply be/src/olap/tablet.cpp:282]: It is necessary to get write lock of `_meta_lock` before exec `Tablet::modify_rowsets`. So, we need not lock `_meta_lock ` here.

File: be/src/olap/compaction.cpp [hunk 2]
```diff
@@ -174,13 +174,17 @@ Status Compaction::construct_input_rowset_readers() {
...
-
     std::lock_guard<std::shared_mutex> wrlock(_tablet->get_header_lock());
-    _tablet->modify_rowsets(output_rowsets, _input_rowsets);
-    _tablet->save_meta();
+    if (_tablet->modify_rowsets(output_rowsets, _input_rowsets, true)) {
+        _tablet->save_meta();
+        return Status::OK();
+    } else {
+        return Status::OLAPInternalError(OLAP_ERR_DELETE_VERSION_ERROR)
+        return ;
```
Comment (wangbo) [inline be/src/olap/compaction.cpp:186]: duplicate return here.

File: be/src/olap/compaction.cpp [hunk 3]
```diff
@@ -114,7 +114,7 @@ Status Compaction::do_compaction_impl(int64_t permits) {
     TRACE("check correctness finished");
 
     // 4. modify rowsets in memory
-    modify_rowsets();
+    RETURN_NOT_OK(modify_rowsets())
```
Comment (morningman) [inline be/src/olap/compaction.cpp:117]:
```suggestion
    RETURN_NOT_OK(modify_rowsets());
```

File: be/src/olap/compaction.cpp [hunk 4]
```diff
@@ -174,13 +174,16 @@ Status Compaction::construct_input_rowset_readers() {
...
 
-void Compaction::modify_rowsets() {
+Status Compaction::modify_rowsets() {
     std::vector<RowsetSharedPtr> output_rowsets;
     output_rowsets.push_back(_output_rowset);
-
     std::lock_guard<std::shared_mutex> wrlock(_tablet->get_header_lock());
-    _tablet->modify_rowsets(output_rowsets, _input_rowsets);
-    _tablet->save_meta();
+    if (_tablet->modify_rowsets(output_rowsets, _input_rowsets, true)) {
```
Comment (morningman) [inline be/src/olap/compaction.cpp:181]: Change to:
```
RETURN_NOT_OK(_tablet->modify_rowsets(output_rowsets, _input_rowsets, true));
_tablet->save_meta();
return Status::OK();
```

File: be/src/olap/tablet.cpp [hunk 5]
```diff
@@ -247,8 +248,9 @@ Status Tablet::add_rowset(RowsetSharedPtr rowset, bool need_persist) {
     return Status::OK();
 }
 
-void Tablet::modify_rowsets(std::vector<RowsetSharedPtr>& to_add,
-                            std::vector<RowsetSharedPtr>& to_delete) {
+bool Tablet::modify_rowsets(std::vector<RowsetSharedPtr>& to_add,
```
Comment (morningman) [inline be/src/olap/tablet.cpp:251]: Change return type to Status

**Review by morningman — COMMENTED** [review]
I think we should still support compaction new tablet when doing alter operation.

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**